doc/source/user/basics.types.rst

```diff
@@ -12,6 +12,146 @@ Array types and conversions between types
+(``int``), unsigned integers (``uint``) floating point (``float``) and
+``complex``. A basic numerical type name combined with a numeric bitsize defines
+a concrete type.  The bitsize is the number of bits that are needed to represent
+a single value in memory. For example, `numpy.float64` is a 64 bit
+floating point data type. Some types, such as `numpy.int` and
+`numpy.intp`, have differing bitsizes, dependent on the platforms
+(e.g. 32-bit vs. 64-bit CPU architectures).  This should be taken into account
+when interfacing with low-level code (such as C or Fortran) where the raw memory
+is addressed.
+
+Data Types for Strings and Bytes
+--------------------------------
+
+In addition to numerical types, NumPy also supports storing unicode strings, via
+`numpy.dtypes.StrDType` (``U`` character code), null-terminated byte sequences via
+`numpy.dtypes.BytesDType` (``S`` character code), and arbitrary byte sequences, via
+`numpy.dtypes.VoidDType` (``V`` character code).
+
+The ``StrDType``, ``BytesDType``, and ``VoidDType`` are *fixed-width* data
+types. They are parameterized by a width, in either bytes or unicode points,
+that a single data element in the array must fit inside. This means that storing
+an array of byte sequences or strings using this dtype requires knowing or
+calculating the sizes of the longest text or byte sequence in advance.
+
+As an example, we can create an array storing the words ``"hello"`` and
+``"world!"``::
+
+  >>> np.array(["hello", "world!"])
+  array(['hello', 'world!'], dtype='<U6')
+
+Here the data type is detected as a unicode string that is a maximum of 6 code
+points long, enough to store both entries without truncation. If we specify a
+shorter or longer data type, the string is either truncated or zero-padded to
+fit in the specified width::
+
+  >>> np.array(["hello", "world!"], dtype="U5")
+  array(['hello', 'world'], dtype='<U5')
+  >>> np.array(["hello", "world!"], dtype="U7")
+  array(['hello', 'world!'], dtype='<U7')
+
+We can see the zero-padding a little more clearly if we use the bytes data
+type and ask NumPy to print out the bytes in the array buffer::
+
+  >>> np.array(["hello", "world"], dtype="S7").tobytes()
+  b'hello\x00\x00world\x00\x00'
+
+Each entry is padded with two extra null bytes. Note however that NumPy cannot
+tell the difference between intentionally stored trailing nulls and padding
+nulls::
+
+  >>> x = [b"hello\0\0", b"world"]
+  >>> a = np.array(x, dtype="S7")
+  >>> print(a[0])
+  b"hello"
+  >>> a[0] == x[0]
+  False
+
+If you need to store and round-trip any trailing null bytes, you will need to
+use an unstructured void data type::
+
+  >>> a = np.array(x, dtype="V7")
+  >>> a
+  array([b'\x68\x65\x6C\x6C\x6F\x00\x00', b'\x77\x6F\x72\x6C\x64\x00\x00'],
+        dtype='|V7')
+  >>> a[0] == np.void(x[0])
+  True
+
+Advanced types, not listed above, are explored in section
+:ref:`structured_arrays`.
+
+Relationship Between NumPy Data Types and C Data Data Types
```

Comment:
You could have a `Reletaionship Between Numpy and Python Data Types here`.

Reply:
That's true, but I'd prefer not to do that right now, if that's OK.

Reply:
I think it is less necessary now - and in any case fine not to do it in this PR.